core/src/HIP/Kokkos_HIP_Instance.hpp

```diff
@@ -128,6 +131,161 @@ class HIPInternal {
 
   HIPInternal() = default;
 
+  // Using HIP API function/objects will be w.r.t. device 0 unless
+  // hipSetDevice(device_id) is called with the correct device_id.
+  // The correct device_id is stored in the variable
+  // HIPInternal::m_hipDev set in HIP::impl_initialize(). In the case
+  // where multiple HIP instances are used, or threads are launched
+  // using non-default HIP execution space after initialization, all HIP
+  // API calls and uses of hipStream_t variables must follow a call to
+  // hipSetDevice(device_id) when an execution space or HIPInternal
+  // object is provided to ensure all computation is done on the correct device.
+
+  // Set the device in to the device stored by this instance for HIP API calls.
+  void set_hip_device() const {
+    verify_is_initialized("set_hip_device");
+    KOKKOS_IMPL_HIP_SAFE_CALL(hipSetDevice(m_hipDev));
+  }
+
+  // Return this instances stream after setting the correct device
+  hipStream_t get_stream() const {
+    set_hip_device();
```

Comment:
Why is it necessary to set the device? We return a stored variable.

Reply:
If I recall correctly, using the stream in the kernel launcher, e.g., 
```
(base_t::get_kernel_func())<<<grid, block, shmem,
                                  hip_instance->get_stream()>>>();
```
required the device to be set. Let me investigate and verify this.

Reply:
If that's the only place, I would prefer if we do something like:
```c++
hip_instance->set_device();
(base_t::get_kernel_func())<<<grid, block, shmem, hip_instance->m_stream>>>();
```

Reply:
For cuda, not setting the correct device causes computation in those kernels to not be done/written on correct device. In HIP there isn't that same issue, at least for the unit test we have. I also did some timings and there seems to be no timing degradation.

But I am still not sure why this isn't an issue for HIP. I know initially running HIP on non-default device but never setting the correct device ID for API calls was not an "issue" in the sense that computation all happened in the correct place, but that the time to execute kernels was very slow. My guess is that HIP allowed each device to access memory on other devices and that was slow?

I'm going to continue to look through the kernel launch impl in Cuda and HIP to try and reason with myself why we would/wouldn't need to set device for Cuda/HIP.

Reply:
I can live with how this is. Is this still a thing to discuss?